### PR TITLE
Group Mask Visibility Sample

### PR DESCRIPTION
**Pull request by patole**
Adding a new sample that demonstrates assigning the appearance.visibility property a group mask instead of a boolean. Each user can independently highlight and select one of 4 buttons, and the results of their highlight and selection remain hidden to everyone else but themselves.

### CODE REVIEW

File: samples/group-mask/README.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
+Shows a text that animates when highlighting or clicking on a cube.
```
Comment (stevenvergenz) [inline samples/group-mask/README.md:1]: I think this readme needs to be updated from the hello world sample.

File: samples/group-mask/README.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+## Setup
+
+* Open a command prompt to this sample's folder and run `npm install`. Keep the command prompt open if you wish to follow the command-oriented instructions that follow.
+* Open the root folder of this repo in VSCode if you wish to follow the VSCode-oriented instructions.
+
+## Build
+
+* Command line: `npm run build`.
+* VSCode: `Shift+Ctrl+B`, then select 'build samples/hello-world'.
```
Comment (stevenvergenz) [inline samples/group-mask/README.md:11]: >hello-world [](start = 53, length = 11)

Here

File: samples/group-mask/README.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+## Run
+
+* Command line: `npm start`.
+* VSCode: Switch to the 'Run' tab (`Ctrl+Shift+D` will open it), select 'Launch hello-world project' from the dropdown at the top, and then `F5` to start it.
```
Comment (stevenvergenz) [inline samples/group-mask/README.md:16]: >hello-world [](start = 80, length = 11)

Here

File: samples/group-mask/README.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+Whether developing an MRE or another kind of app, an efficient dev/test loop is essential. Devs familiar with making browser-based apps will be familiar with webpack's notion of "hot reload", that is, automatically applying changes as they're made without the need to explicitly stop/rebuild/restart your app. NodeJS apps can do this too.
+
+This setup requires launching the app from a terminal. VSCode has a built-in terminal, or you can open a separate command prompt.
+
+### Start the MRE with hot-reload enabled
+
+1. In the terminal, in this project's folder, run: `npm run debug-watch`. This will build and start the MRE. The `debug-watch` task continues to run in the background, watching for code changes. It will rebuild and restart the app whenever files are modified.
+2. In VSCode, press `Ctrl+Shift+D` to open the 'Run' tab, select 'Attach to running project' from the drop down at the top, then press `F5` to attach the VSCode debugger. This step isn't required, but allows you to set breakpoints and debug MRE execution.
+
+### See hot-reload in action
```
Comment (stevenvergenz) [inline samples/group-mask/README.md:47]: >See hot-reload in action [](start = 4, length = 24)

And this section

File: samples/group-mask/src/groupMaskManager.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	CLICK = ButtonMask.CLICK_A,
+	LEAVE = ButtonMask.DEFAULT_B,
+	LEAVE_HOVER = ButtonMask.HOVER_B
+}
+
+export default class GroupMaskManager {
+	private masks: MRE.GroupMask[];
+	private readonly PREFIX = 'MASK_';
+
+	public constructor(app: GroupMask) {
```
Comment (stevenvergenz) [inline samples/group-mask/src/groupMaskManager.ts:42]: Please rename this import, it's confusing with MRE.GroupMask.

File: samples/group-mask/src/app.ts [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+	constructor(public context: MRE.Context, private baseUrl: string) {
+		this.context.onStarted(() => this.started());
+
+		this.groupMaskManager = new GroupMaskManager(this);
+	}
+
+	private started() {
+		this.assets = new MRE.AssetContainer(this.context);
+		const whiteMaterial = this.assets.createMaterial("whiteMat", {
+			color: Color4.FromColor3(Color3.White(), .7),
```
Comment (stevenvergenz) [inline samples/group-mask/src/app.ts:45]: >Color4.FromColor3(Color3.White(), .7) [](start = 10, length = 37)

For black and white, it's probably shorter to just supply RGBA.

File: samples/group-mask/src/app.ts [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+							size:{x:.4, y:.4, z:.4}
+						},
+						layer: MRE.CollisionLayer.Default
+					}
+				},
+			});
+
+			const buttonBehavior = baseButton.setBehavior(MRE.ButtonBehavior);
+
+			buttonBehavior.onClick(user => {
```
Comment (stevenvergenz) [inline samples/group-mask/src/app.ts:81]: >.onClick [](start = 17, length = 8)

These three handlers can be chained if desired: `buttonBehavior.onClick(...).onHover(...)`

File: samples/group-mask/src/app.ts [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+					},
+				},
+			});
+
+
+			highlightedButton.appearance.enabled = this.groupMaskManager.getButtonMask(ButtonMask.HOVER_A + i); 
+			this.highlightCubes.push(highlightedButton);
+
+			//Load a glTF model for the selected state
+			const selectedCube = MRE.Actor.CreateFromGltf(this.assets, {
```
Comment (stevenvergenz) [inline samples/group-mask/src/app.ts:117]: >CreateFromGltf [](start = 34, length = 14)

Don't call `CreateFromGltf` in a loop, you're loading the glTF file four times. Instead call `loadGltf` once at the top, and `CreateFromPrefab` to spawn the cube. Though this is a common-enough mistake that I wonder if we need to change the API.

File: samples/group-mask/package.json [hunk 9]
```diff
@@ -0,0 +1,46 @@
+{
+  "name": "hello-world",
```
Comment (tombuMS) [inline samples/group-mask/package.json:2]: >hello-world [](start = 11, length = 11)

Update name

**Comment by tombuMS** [comment]
>   "name": "hello-world",

Update name

---
Refers to: samples/group-mask/package-lock.json:2 in 997b0a4. [](commit_id = 997b0a401273558dc50cbb0ad3b3f1e29dda993b, deletion_comment = False)

File: samples/group-mask/src/app.ts (hunk 7 above)
Comment (patole) [reply samples/group-mask/src/app.ts:81]: For me that really reduces comprehension and readability for negligible gain.

File: samples/group-mask/src/app.ts (hunk 8 above)
Comment (patole) [reply samples/group-mask/src/app.ts:117]: The multiple overlapping Actor/Asset creation APIs are a pain point for sure

**Pull request merged**